### DISCUSSION TITLE
How does one create a new patch?

### DISCUSSION THREAD

**Discussion by RangerMauve** [bromite/bromite#1908] [Builds]
What are the steps that developers take to develop new patches?

I'm currently working on forking bromite to integrate peer to peer protocols for use in mesh networks in South Africa via the [Agregore Browser](https://agregore.mauve.moe/) and I'm not sure how to actually go about this.

I've managed to get some build scripts to apply the bromite patches to chromium's source code, but I'm not sure if there are guides anywhere for how to develop new patches and additions.

We've got some budget to figure this out, so it'd be nice to pay a dev for a couple of hours of their time to explain their workflow. I'd also love to take what we learn and write a wiki entry to give people more info on how they can contribute code changes to the browser.

**Comment by csagan5** [comment]
> I've managed to get some build scripts to apply the bromite patches to chromium's source code, but I'm not sure if there are guides anywhere for how to develop new patches and additions.

It is not any different from building and patching/modifying Chromium; you should start from building Chromium (there is plenty of documentation about that) and then it is not much different than developing a regular Android app.

> We've got some budget to figure this out, so it'd be nice to pay a dev for a couple of hours of their time to explain their workflow.

Sorry, that is not possible; being proficient with Chromium builds requires continuous support (not a couple of hours). More build tooling might be released in future but providing support for custom builds is not something I can commit to.

**Reply by RangerMauve** [reply]
Thank you for the response! I already had some experience compiling chrome which helped, and put together a small script that would checkout the correct version of chromium, copy over the `gn.args` and apply the patches for bromite by going through the list. 😁

Hopefully this could help some folks get started in the future if they're not sure how to use the stuff in the repo.

**Reply by csagan5** [reply]
There are some unofficial ones you could use as starting point:
* https://github.com/uazo/bromite-buildtools
* https://github.com/nikolowry/bromite-builder

There is also this re-branded fork that has build tooling but they do not participate in community nor contribute back any patch:
* https://gitlab.e.foundation/e/os/browser

If you are working on IPFS I would be interested in merging a patch about that.

**Reply by RangerMauve** [reply]
Awesome, thank you for the resources.

I'd love to PR the patch once it's working. 😁 I'm taking a similar patch-based workflow to what you have in Bromite so it should be pretty easy to incorporate.

At the moment we've got a golang based binary that we've compiled to run on Android [here](https://github.com/AgregoreWeb/agregore-ipfs-daemon). I'm in the process of figuring out how to actually integrate it into the codebase. I had initially put it under `res/raw` and tried extracting it to the files folder and executing it, but that method doesn't work in Android Q, so I'm looking at how I can put it int he `lib/armabi-v7a` folder since it's executable. The work is being tracked here: https://github.com/AgregoreWeb/agregore-mobile/issues/6

If you have any tips on other ways of integrating the gateway with the chromium codebase I'd very much appreciate it. I'm going to try to just monkey patch the binary into the final APK after it's compiled, but longer term it'd be nice to have the [gomobile](https://pkg.go.dev/golang.org/x/mobile/cmd/gomobile) compilation somehow be part of the GN setup.

Something that might also be of interest to you that we're planning on developing in the next month or two is [Wifi Direct](https://developer.android.com/training/connect-devices-wirelessly/wifi-direct) support so that two people with the app can do IPFS based p2p directly with each other even without the internet or an access point. The UX is going to be a button somewhere in the menu bar which will give connection status indicators and trigger an auto-discovery/connect when users click it.

The full scope of the work we're doing is here: https://github.com/ipfs/devgrants/blob/master/open-grants/open-proposal-agregore-mobile.md

We're also planning on adding more p2p protocols in the future and working with community mesh network projects. If stuff works out with this maybe we could apply to a future grant together if there's functionality that'd be useful to Bromite as well. 😁

**Reply by csagan5** [reply]
> At the moment we've got a golang based binary that we've compiled to run on Android [here](https://github.com/AgregoreWeb/agregore-ipfs-daemon). I'm in the process of figuring out how to actually integrate it into the codebase. I had initially put it under `res/raw` and tried extracting it to the files folder and executing it, but that method doesn't work in Android Q, so I'm looking at how I can put it int he `lib/armabi-v7a` folder since it's executable. The work is being tracked here: [AgregoreWeb/agregore-mobile#6](https://github.com/AgregoreWeb/agregore-mobile/issues/6)

I think Brave has a better approach:
* https://brave.com/ipfs-support/
* https://github.com/brave/brave-core/tree/master/components/ipfs
Some time ago I wanted to try adding this but I never got around to it; if you want to submit a patch for Bromite I am afraid it would have to be C++/Java.

> We're also planning on adding more p2p protocols

Out of curiosity, which ones are you looking at?

**Reply by RangerMauve** [reply]
> I think Brave has a better approach

Brave actually takes a similar approach to us and [downloads a binary at runtime](https://github.com/brave/brave-core/blob/master/components/ipfs/brave_ipfs_client_updater.h#L88) which contains the go-ipfs code. I'm not sure what their approach is for mobile, but it's likely they're just reaching an external gateway (which is what most other browsers with IPFS support do) which won't work well with the fully-local and offline-first use case that we're doing in Agregore. 😅

There's also the option to use js-ipfs, but the reliability and performance of that is kinda garbage at the moment. I was actually using that in Agregore Desktop up until recently and it just didn't deliver good enough UX to be usable. :/ We're also going to be doing a bunch of work in our local gateway to optimize the network usage to improve battery life.

I'd really like to use something that's plain C++/Java but sadly Golang has the most reliable implementation at the moment. 😩 FWIW the folks at Igalia are working on making an approach using `registerProtocolHandler` inside web extensions, but I'm still not convinced that js-ipfs will be able to give reliable enough performance compared to go-ipfs any time soon.

Regarding protocols, we're likely going to support everything that's supported in Agregore Desktop, but right now my priorities are [BitTorrent](http://bittorrent.org/beps/bep_0000.html) and [Hypercore](https://hypercore-protocol.org/), but I'm not sure how to fund work on that yet.

**Reply by csagan5** [reply]
> > I think Brave has a better approach
> 
> Brave actually takes a similar approach to us and [downloads a binary at runtime](https://github.com/brave/brave-core/blob/master/components/ipfs/brave_ipfs_client_updater.h#L88) which contains the go-ipfs code. I'm not sure what their approach is for mobile, but it's likely they're just reaching an external gateway (which is what most other browsers with IPFS support do) which won't work well with the fully-local and offline-first use case that we're doing in Agregore. sweat_smile

Fascinating, thanks for the information; and the goal of being fully-local and offline-first is indeed a great one. :+1: 

> I'd really like to use something that's plain C++/Java but sadly Golang has the most reliable implementation at the moment.

I was mentioning it because it would work best (from user and patch maintainers' perspective) with the components structure of Chromium, let me know how it goes with the Go external binary. What will you use for IPC? Mojo or something else?

> Regarding protocols, we're likely going to support everything that's supported in Agregore Desktop, but right now my priorities are [BitTorrent](http://bittorrent.org/beps/bep_0000.html) and [Hypercore](https://hypercore-protocol.org/), but I'm not sure how to fund work on that yet.

Very interesting, feel free to post here on the discussions if you need help with the Bromite patches.

**Reply by RangerMauve** [reply]
> let me know how it goes with the Go external binary. What will you use for IPC? Mojo or something else?

Right now we're planning to do HTTP over a UNIX domain socket. The C++ protocol handler will handle incoming requests from the browser, open HTTP requests to the go daemon and pipe everything back to the response. This is similar to the approach we have in Agregore Desktop in that the URLs for p2p stuff easily map to paths on a gateway.

Right now even before dealing with golang and the such, I just wish I could find where in the code I could get it to copy a file into the final APK. Seems like it's generating a bunch of `.jar` files and merging them together, but I'm still reading through the code to figure out how it works. 😭 The build codebase is so huge and fractured.

Also, would you be down for a small PR to the README to add details on how to use the patches?

**Reply by RangerMauve** [reply]
@csagan5 We've got an initial version of our IPFS/IPNS protocol handlers working:

This patch adds code to spawn an IPFS daemon on activity start: https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-IPFS-Daemon.patch

Here's the source code for the daemon: https://github.com/AgregoreWeb/agregore-ipfs-daemon/

We add it to the codebase by adding a compiled AAR file to the build tree.

And this patch adds the custom protocol handlers which redirect to the daemon:
https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch

If you're not interested in running a local node, you could use these patches to point to a public IPFS gateway, though note that this will only give you the ability to read data and not publish it, plus it requires you to trust the gateway to not do sussy stuff with the data first.

Do either of these thing seem relevant to add to Bromite? I'd be down to do a PR if so.

**Reply by uazo** [reply]
can I ask you a question?
reading the patch it seems to me that for chromium the only part really needed is [this]( https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch) because the integration actually seems to me to be a redirect to localhost:8080, correct?
the rest is the start of the gateway which, however, could be a stand-alone application launched if necessary by the user, correct?
thanks

**Reply by csagan5** [reply]
> We add it to the codebase by adding a compiled AAR file to the build tree.
> 
> And this patch adds the custom protocol handlers which redirect to the daemon: https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch
> 
> If you're not interested in running a local node, you could use these patches to point to a public IPFS gateway, though note that this will only give you the ability to read data and not publish it, plus it requires you to trust the gateway to not do sussy stuff with the data first.
> 

It would need to support functionality to turn on/off this daemon and not run it all the times, so that users that don't want to participate in the experiment can leave it disabled by default.

Also about the approach: will the Android security scanner flag this as a malicious behaviour? Perhaps a companion APK would be a better way to distribute the daemon? (Assuming that it can be made Android-native, then with its own settings etc)

Any other concern (e.g. IPFS name resolution) that we should think about?

> Do either of these thing seem relevant to add to Bromite? I'd be down to do a PR if so.

Let's discuss it a bit first if you don't mind (to avoid wasting your time), then I will tell you to go ahead.

> the only part really needed is [this](https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch) because the integration actually seems to me to be a redirect to localhost:8080, correct?
> the rest is the start of the gateway which, however, could be a stand-alone application launched if necessary by the user, correct?

I think so. Using HTTP is probably better than some custom IPC like mojo so that multiple daemons can be used (even if there is practically just one available right now).

**Reply by RangerMauve** [reply]
> can I ask you a question? reading the patch it seems to me that for chromium the only part really needed is [this](https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch) because the integration actually seems to me to be a redirect to localhost:8080, correct? the rest is the start of the gateway which, however, could be a stand-alone application launched if necessary by the user, correct? thanks

Yup exactly. If you want you can skip the patch for adding a localhost and have it redirect to something like `https://ipfs.io/ipfs/`.

We're using a local gateway because we want to be able to write data to IPFS with protocol handlers and we want stuff to work on local networks without the internet.

**Reply by RangerMauve** [reply]
> It would need to support functionality to turn on/off this daemon and not run it all the times, so that users that don't want to participate in the experiment can leave it disabled by default.

Yeah, absolutely agreed. I was thinking of finding a way to enable it the first time you access an `ipfs://` or `ipns://` URL. We could probably add a button in the app menu to shut it down / turn it on, too. Currently it shuts down once the activity closes.


> Also about the approach: will the Android security scanner flag this as a malicious behaviour? Perhaps a companion APK would be a better way to distribute the daemon? (Assuming that it can be made Android-native, then with its own settings etc)

Honestly not sure. It's just a go library embedded within the app, so I'd assume it'd be okay. The gateway isn't a separate process or anything, just invoked via JNI. A companion APK isn't something I'd like to work on since it adds a bunch of complexity around versioning and initialization UX. 😅

> Any other concern (e.g. IPFS name resolution) that we should think about?

Right now we've hardcoded to use `1.1.1.1` for DNS queries for when you use a DNS based IPNS name like `ipns://ipfs.io`, not sure if that's an issue. Else all the traffic is encrypted so it shouldn't be too bad for privacy.

> I think so. Using HTTP is probably better than some custom IPC like mojo so that multiple daemons can be used (even if there is practically just one available right now).

Yeah, I was also thinking that Unix domain sockets might be good for security, but then we couldn't use this redirect based approach. 😅

Also contemplating randomizing the port or using something other than 8080 since there's a lot of conflicts for that port number.

**Reply by uazo** [reply]
> If you want you can skip the patch for adding a localhost and have it redirect to something like https://ipfs.io/ipfs/

no, that's not why, maybe I suggest you take a look at [Block-qjz9zk-or-trk-requests.patch](https://github.com/bromite/bromite/blob/master/build/patches/Block-qjz9zk-or-trk-requests.patch), if you haven't already.

> It would need to support functionality to turn on/off this daemon and not run it all the times,

I think you can take a look at [this](https://github.com/bromite/bromite/blob/f819874002a357e93e1b13c5c32d0f11013b1183/build/patches/Enable-share-intent.patch#L239) maybe it can be useful to you

>> Also about the approach: will the Android security scanner flag this as a malicious behaviour?
>
> Honestly not sure.

if I may have my say, this worries me more: **though IPFS is not production-ready yet** from your upstream.

> It's just a go library embedded within the app, so I'd assume it'd be okay

however, although I am not able to read the go code carefully, I am concerned that the daemon runs in the same context as chromium, with the same rights.
a possible bug of the daemon would allow to enter the browser folder, which is not protected, and considering that (it seems to me) that the cache is on filesystems, theoretically I could also have executables, it seems to me quite dangerous.

for what is perhaps better to be a companion APK, at least in a first phase

**Reply by RangerMauve** [reply]
> no, that's not why, maybe I suggest you take a look at [Block-qjz9zk-or-trk-requests.patch](https://github.com/bromite/bromite/blob/master/build/patches/Block-qjz9zk-or-trk-requests.patch), if you haven't already.

That's super useful, I think we actually just ran into some issues while testing which this might help us with. Thank you!

> if I may have my say, this worries me more: **though IPFS is not production-ready yet** from your upstream.

Would you mind linking me to where you're reading that? IPFS might not be 100% stable, but it's stable enough for people to build applications on top of and has a fairly large userbase from what I've seen.

It's probably more experimental than HTTPS, though. 🤪
 
> however, although I am not able to read the go code carefully, I am concerned that the daemon runs in the same context as chromium, with the same rights. a possible bug of the daemon would allow to enter the browser folder, which is not protected, and considering that (it seems to me) that the cache is on filesystems, theoretically I could also have executables, it seems to me quite dangerous.

I've got it isolated in it's own process inside a BoundService at the moment. Would adding more strict restrictions to that process help with your worries?

Regarding executables, Android 10 and up doesn't allow executables within writable directories anymore, and it seems the only way to have executables at all is via libraries that are made for JNI.

I could see how a companion APK would be more favorable. Sadly, that's outside the scope of what I have time to maintain / contribute. 😅 If folks are particularly interested in making that happen I could get them in touch with some Protocol Labs folks and could probably help secure funding to make it happen.

**Reply by uazo** [reply]
> Would you mind linking me to where you're reading that

from https://github.com/ipfs/go-ipfs/tree/v0.12.2#security-issues

> Sadly, that's outside the scope of what I have time to maintain / contribute.

mine was just a note.
free to do and propose what you want of course!
